File: src/main/java/org/dementhium/action/Action.java
package org.dementhium.action;

import org.dementhium.event.Tickable;
import org.dementhium.model.Mob;

/**
 *
 * @author 'Mystic Flow <Steven@rune-server.org>
 *
 */
public abstract class Action extends Tickable {

	protected Mob mob;
	
	public Action(int cycles) {
		super(cycles);
	}
	
	public void setEntity(Mob mob) {
		this.mob = mob;
	}
	
	@Override
	public void stop() {
		super.stop();//stops the main tickable
		mob.getActionManager().appendAction(mob.getActionManager().poll());
	}
}

File: src/main/java/org/dementhium/action/Action.kt
package org.dementhium.action

import org.dementhium.event.Tickable
import org.dementhium.model.Mob
import kotlin.properties.Delegates

abstract class Action(cycles: Int) : Tickable(cycles) {

    companion object {
        val NoAction = object: Action(-1) {
            override fun execute() { }
        }
    }

    protected var mob: Mob by Delegates.notNull()

    fun setEntity(mob: Mob) { this.mob = mob }

    override fun stop() {
        super.stop()
        mob.actionManager.appendAction(mob.actionManager.poll())
    }
}
